### ISSUE TITLE
Bug: Somehow a TupleLiteral ends up as TypeNode generic argument

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#10056]
The following code produces a weird error:
```cr
module Indexable(T)
  def join(separator = "")
    previous_def
  end
end

Array(Int32).new.join
```

```
In /usr/share/crystal/src/indexable.cr:332:21

 332 | {% elsif String < T %}
                       ^
Error: TypeNode#< expects TypeNode, not TupleLiteral
```

I couldn't find a more reduced and isolated example.

**Comment by asterite** [comment]
Well, in the case of a Tuple, T is a tuple

**Comment by straight-shoota** [comment]
I don't follow. How can `T` be a tuple?

**Comment by asterite** [comment]
Try defining a foo method on a tuple and printing T inside a macro. However, I don't know why T is a tuple in the Indexable case. Probably something is not working well.

**Comment by asterite** [comment]
It seems the T isn't well computed when using super or previous def

**Comment by HertzDevil** [comment]
Seems to be fixed since 0.36
